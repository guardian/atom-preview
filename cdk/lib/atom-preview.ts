import { GuEc2App } from '@guardian/cdk';
import { AccessScope } from '@guardian/cdk/lib/constants';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuCname } from '@guardian/cdk/lib/constructs/dns';
import type { App } from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import { InstanceClass, InstanceSize, InstanceType } from 'aws-cdk-lib/aws-ec2';
import {bootstrapCommand} from "@guardian/cdk/lib/bin/commands/bootstrap";
import {GuAllowPolicy} from "@guardian/cdk/lib/constructs/iam";

interface AtomProps extends GuStackProps {
	domainName: string;
	minimumInstances: number;
}
export class AtomPreview extends GuStack {
	constructor(scope: App, id: string, props: AtomProps) {
		super(scope, id, props);
		const { loadBalancer } = new GuEc2App(this, {
			applicationPort: 9000,
			app: 'atom-preview',
			access: { scope: AccessScope.PUBLIC },
			instanceType: InstanceType.of(InstanceClass.T4G, InstanceSize.SMALL),
			certificateProps: {
				domainName: props.domainName,
			},
			monitoringConfiguration: {
				noMonitoring: true,
			},
			applicationLogging: {
				enabled: true
			},
			userData: {
				distributable: {
					fileName: 'atom-preview.deb',
					executionStatement: `dpkg -i /atom-preview/atom-preview.deb`,
				},
			},
			scaling: {
				minimumInstances: props.minimumInstances,
			},
			roleConfiguration: {
				additionalPolicies: [
					new GuAllowPolicy(this, "PanDomainPolicy", {
						resources: ["arn:aws:s3:::pan-domain-auth-settings/*"],
						actions: ["s3:GetObject"],
					}),
				]
			}
		});
		new GuCname(this, 'dns', {
			app: 'atom-preview',
			domainName: props.domainName,
			resourceRecord: loadBalancer.loadBalancerDnsName,
			ttl: Duration.hours(1),
		});
	}
}
