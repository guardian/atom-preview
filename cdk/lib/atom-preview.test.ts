import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { AtomPreview } from './atom-preview';

describe('The AtomPreview stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new AtomPreview(app, 'AtomPreview', {
			stack: 'flexible',
			stage: 'PROD',
			minimumInstances: 1,
			domainName: 'atom-preview.gutools.co.uk',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
