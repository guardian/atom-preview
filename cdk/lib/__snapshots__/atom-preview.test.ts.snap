// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The AtomPreview stack matches the snapshot 1`] = `
Object {
  "Outputs": Object {
    "LoadBalancerAtompreviewDnsName": Object {
      "Description": "DNS entry for LoadBalancerAtompreview",
      "Value": Object {
        "Fn::GetAtt": Array [
          "LoadBalancerAtompreview923A2964",
          "DNSName",
        ],
      },
    },
  },
  "Parameters": Object {
    "AMIAtompreview": Object {
      "Description": "Amazon Machine Image ID for the app atom-preview. Use this in conjunction with AMIgo to keep AMIs up to date.",
      "Type": "AWS::EC2::Image::Id",
    },
    "DistributionBucketName": Object {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "LoggingStreamName": Object {
      "Default": "/account/services/logging.stream.name",
      "Description": "SSM parameter containing the Name (not ARN) on the kinesis stream",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "VpcId": Object {
      "Default": "/account/vpc/primary/id",
      "Description": "Virtual Private Cloud to run EC2 instances within. Should NOT be the account default VPC.",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::VPC::Id>",
    },
    "atompreviewPrivateSubnets": Object {
      "Default": "/account/vpc/primary/subnets/private",
      "Description": "A list of private subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
    "atompreviewPublicSubnets": Object {
      "Default": "/account/vpc/primary/subnets/public",
      "Description": "A list of public subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
  },
  "Resources": Object {
    "AutoScalingGroupAtompreviewASGE9B8D8AD": Object {
      "Properties": Object {
        "HealthCheckGracePeriod": 120,
        "HealthCheckType": "ELB",
        "LaunchConfigurationName": Object {
          "Ref": "AutoScalingGroupAtompreviewLaunchConfig02DE693A",
        },
        "MaxSize": "2",
        "MinSize": "1",
        "Tags": Array [
          Object {
            "Key": "App",
            "PropagateAtLaunch": true,
            "Value": "atom-preview",
          },
          Object {
            "Key": "gu:cdk:pattern-name",
            "PropagateAtLaunch": true,
            "Value": "GuEc2App",
          },
          Object {
            "Key": "gu:cdk:version",
            "PropagateAtLaunch": true,
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "PropagateAtLaunch": true,
            "Value": "guardian/atom-preview",
          },
          Object {
            "Key": "LogKinesisStreamName",
            "PropagateAtLaunch": true,
            "Value": Object {
              "Ref": "LoggingStreamName",
            },
          },
          Object {
            "Key": "Name",
            "PropagateAtLaunch": true,
            "Value": "AtomPreview/AutoScalingGroupAtompreview",
          },
          Object {
            "Key": "Stack",
            "PropagateAtLaunch": true,
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "PropagateAtLaunch": true,
            "Value": "PROD",
          },
          Object {
            "Key": "SystemdUnit",
            "PropagateAtLaunch": true,
            "Value": "atom-preview.service",
          },
        ],
        "TargetGroupARNs": Array [
          Object {
            "Ref": "TargetGroupAtompreviewB7AA27C8",
          },
        ],
        "VPCZoneIdentifier": Object {
          "Ref": "atompreviewPrivateSubnets",
        },
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
    },
    "AutoScalingGroupAtompreviewInstanceProfile2E5A9B96": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "InstanceRoleAtompreview4AABC81B",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "AutoScalingGroupAtompreviewLaunchConfig02DE693A": Object {
      "DependsOn": Array [
        "InstanceRoleAtompreview4AABC81B",
      ],
      "Properties": Object {
        "IamInstanceProfile": Object {
          "Ref": "AutoScalingGroupAtompreviewInstanceProfile2E5A9B96",
        },
        "ImageId": Object {
          "Ref": "AMIAtompreview",
        },
        "InstanceType": "t4g.small",
        "MetadataOptions": Object {
          "HttpTokens": "required",
        },
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "GuHttpsEgressSecurityGroupAtompreview72923F0F",
              "GroupId",
            ],
          },
          Object {
            "Fn::GetAtt": Array [
              "WazuhSecurityGroup",
              "GroupId",
            ],
          },
        ],
        "UserData": Object {
          "Fn::Base64": Object {
            "Fn::Join": Array [
              "",
              Array [
                "#!/bin/bash
mkdir -p $(dirname '/atom-preview/atom-preview.deb')
aws s3 cp 's3://",
                Object {
                  "Ref": "DistributionBucketName",
                },
                "/flexible/PROD/atom-preview/atom-preview.deb' '/atom-preview/atom-preview.deb'
dpkg -i /atom-preview/atom-preview.deb",
              ],
            ],
          },
        },
      },
      "Type": "AWS::AutoScaling::LaunchConfiguration",
    },
    "CertificateAtompreview18C38C6D": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "DomainName": "atom-preview.gutools.co.uk",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "atom-preview",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/atom-preview",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
      "UpdateReplacePolicy": "Retain",
    },
    "DescribeEC2PolicyFF5F9295": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "autoscaling:DescribeAutoScalingInstances",
                "autoscaling:DescribeAutoScalingGroups",
                "ec2:DescribeTags",
                "ec2:DescribeInstances",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "describe-ec2-policy",
        "Roles": Array [
          Object {
            "Ref": "InstanceRoleAtompreview4AABC81B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GetDistributablePolicyAtompreview7FDF318F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:s3:::",
                    Object {
                      "Ref": "DistributionBucketName",
                    },
                    "/flexible/PROD/atom-preview/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetDistributablePolicyAtompreview7FDF318F",
        "Roles": Array [
          Object {
            "Ref": "InstanceRoleAtompreview4AABC81B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GuHttpsEgressSecurityGroupAtompreview72923F0F": Object {
      "Properties": Object {
        "GroupDescription": "Allow all outbound HTTPS traffic",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound HTTPS traffic",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "atom-preview",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/atom-preview",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "VpcId": Object {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "GuHttpsEgressSecurityGroupAtompreviewfromAtomPreviewLoadBalancerAtompreviewSecurityGroupD74F875190005FB9F28A": Object {
      "Properties": Object {
        "Description": "Load balancer to target",
        "FromPort": 9000,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "GuHttpsEgressSecurityGroupAtompreview72923F0F",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "LoadBalancerAtompreviewSecurityGroupF8A8C6B0",
            "GroupId",
          ],
        },
        "ToPort": 9000,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "GuLogShippingPolicy981BFE5A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:Describe*",
                "kinesis:Put*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:kinesis:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":stream/",
                    Object {
                      "Ref": "LoggingStreamName",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GuLogShippingPolicy981BFE5A",
        "Roles": Array [
          Object {
            "Ref": "InstanceRoleAtompreview4AABC81B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "InstanceRoleAtompreview4AABC81B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Path": "/",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "atom-preview",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/atom-preview",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ListenerAtompreview8FD5C6B0": Object {
      "Properties": Object {
        "Certificates": Array [
          Object {
            "CertificateArn": Object {
              "Ref": "CertificateAtompreview18C38C6D",
            },
          },
        ],
        "DefaultActions": Array [
          Object {
            "TargetGroupArn": Object {
              "Ref": "TargetGroupAtompreviewB7AA27C8",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": Object {
          "Ref": "LoadBalancerAtompreview923A2964",
        },
        "Port": 443,
        "Protocol": "HTTPS",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "LoadBalancerAtompreview923A2964": Object {
      "Properties": Object {
        "LoadBalancerAttributes": Array [
          Object {
            "Key": "deletion_protection.enabled",
            "Value": "true",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "LoadBalancerAtompreviewSecurityGroupF8A8C6B0",
              "GroupId",
            ],
          },
        ],
        "Subnets": Object {
          "Ref": "atompreviewPublicSubnets",
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "atom-preview",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/atom-preview",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "LoadBalancerAtompreviewSecurityGroupF8A8C6B0": Object {
      "Properties": Object {
        "GroupDescription": "Automatically created Security Group for ELB AtomPreviewLoadBalancerAtompreview95F81FCB",
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 443",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "atom-preview",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/atom-preview",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "VpcId": Object {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "LoadBalancerAtompreviewSecurityGrouptoAtomPreviewGuHttpsEgressSecurityGroupAtompreview9DA18F21900087C11F36": Object {
      "Properties": Object {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "GuHttpsEgressSecurityGroupAtompreview72923F0F",
            "GroupId",
          ],
        },
        "FromPort": 9000,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "LoadBalancerAtompreviewSecurityGroupF8A8C6B0",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": 9000,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "PanDomainPolicyB981546E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::pan-domain-auth-settings/*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PanDomainPolicyB981546E",
        "Roles": Array [
          Object {
            "Ref": "InstanceRoleAtompreview4AABC81B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ParameterStoreReadAtompreviewEC146574": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/flexible/atom-preview",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/flexible/atom-preview/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "parameter-store-read-policy",
        "Roles": Array [
          Object {
            "Ref": "InstanceRoleAtompreview4AABC81B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SSMRunCommandPolicy244E1613": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ec2messages:AcknowledgeMessage",
                "ec2messages:DeleteMessage",
                "ec2messages:FailMessage",
                "ec2messages:GetEndpoint",
                "ec2messages:GetMessages",
                "ec2messages:SendReply",
                "ssm:UpdateInstanceInformation",
                "ssm:ListInstanceAssociations",
                "ssm:DescribeInstanceProperties",
                "ssm:DescribeDocumentParameters",
                "ssmmessages:CreateControlChannel",
                "ssmmessages:CreateDataChannel",
                "ssmmessages:OpenControlChannel",
                "ssmmessages:OpenDataChannel",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ssm-run-command-policy",
        "Roles": Array [
          Object {
            "Ref": "InstanceRoleAtompreview4AABC81B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TargetGroupAtompreviewB7AA27C8": Object {
      "Properties": Object {
        "HealthCheckIntervalSeconds": 10,
        "HealthCheckPath": "/healthcheck",
        "HealthCheckProtocol": "HTTP",
        "HealthCheckTimeoutSeconds": 5,
        "HealthyThresholdCount": 5,
        "Port": 9000,
        "Protocol": "HTTP",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "atom-preview",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/atom-preview",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "TargetGroupAttributes": Array [
          Object {
            "Key": "deregistration_delay.timeout_seconds",
            "Value": "30",
          },
          Object {
            "Key": "stickiness.enabled",
            "Value": "false",
          },
        ],
        "TargetType": "instance",
        "UnhealthyThresholdCount": 2,
        "VpcId": Object {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "WazuhSecurityGroup": Object {
      "Properties": Object {
        "GroupDescription": "Allow outbound traffic from wazuh agent to manager",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Wazuh event logging",
            "FromPort": 1514,
            "IpProtocol": "tcp",
            "ToPort": 1514,
          },
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Wazuh agent registration",
            "FromPort": 1515,
            "IpProtocol": "tcp",
            "ToPort": 1515,
          },
        ],
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/atom-preview",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "VpcId": Object {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "dns": Object {
      "Properties": Object {
        "Name": "atom-preview.gutools.co.uk",
        "RecordType": "CNAME",
        "ResourceRecords": Array [
          Object {
            "Fn::GetAtt": Array [
              "LoadBalancerAtompreview923A2964",
              "DNSName",
            ],
          },
        ],
        "Stage": "PROD",
        "TTL": 3600,
      },
      "Type": "Guardian::DNS::RecordSet",
    },
  },
}
`;
